DOCS-2523 - powershell scripts for Azure App Service Extension (#11766)
DataDog/documentation: content/en/serverless/azure_app_services.md

Learn more about [custom metrics][17].

## Extension management with Powershell

> Datadog provides scripts with which you can update or install the Azure App Service Extension using Powershell. Scripted extension management enables you to [update extensions in bulk by resource group](#powershell-resource-group) and to [designate the installation of specific versions of the site extension](#powershell-specific-version). You can also use scripts to programatically add the extension in CI/CD pipelines, as well as discover and update extensions that are already installed.

Reviewer: ```suggestion
Datadog provides scripts to update or install the Azure App Service Extension using Powershell. Scripted extension management enables you to [update extensions in bulk by resource group](#powershell-resource-group) and [designate the installation of specific versions of the site extension](#powershell-specific-version). You can also use scripts to programmatically add the extension in CI/CD pipelines, as well as discover and update extensions that are already installed.
```

### Prerequisites

- The [Azure CLI][18] or [Azure Cloud Shell][19].
- Azure App Service [user-scope credentials][20]. If you do not already have credentials, go to your [Azure portal][21] and access the App Service instance (Web App or Function App). Navigate to **Deployment** > **Deployment Center** to create or retrieve your user-scope credentials.

### Installing the extension for the first time {#powershell-first-time}

The install script adds the latest version of the extension to an Azure Web App or Azure Function App. This occurs on a per-instance basis, rather than at a resource group level.

1. Open the Azure CLI or Azure Cloud Shell.
2. Download the installation script using the following command:

    ```
    Invoke-WebRequest -Uri "https://raw.githubusercontent.com/DataDog/datadog-aas-extension/master/management-scripts/extension/install-latest-extension.ps1" -OutFile "install-latest-extension.ps1"
    ```

> 3. Run the following command, passing in required arguments and any others as needed.

Reviewer: ```suggestion
3. Run the following command, passing in required and optional arguments as needed.
```

The update script applies to an entire resource group. This script updates every instance (Web App or Function App) that has the extension installed. App Service instances that do not have the Datadog extension installed are not affected.

1. Open the Azure CLI or Azure Cloud Shell.
2. Download the installation script using the following command:

    ```
    $baseUri="https://raw.githubusercontent.com/DataDog/datadog-aas-extension/master/management-scripts/extension"; Invoke-WebRequest -Uri "$baseUri/update-all-site-extensions.ps1" -OutFile "update-all-site-extensions.ps1"; Invoke-WebRequest -Uri "$baseUri/install-latest-extension.ps1" -OutFile "install-latest-extension.ps1"
    ```

3. Run the following command. All arguments are required.

    ```
    .\update-all-site-extensions.ps1 -SubscriptionId <SUBSCRIPTION_ID> -ResourceGroup <RESOURCE_GROUP_NAME> -Username <USERNAME> -Password <PASSWORD>
    ```

### Installing a specific version of the extension {#powershell-specific-version}

The Azure App Service UI does not support the ability to install a specific version of an extension. You may do this with the install or update script.

> #### Install specific version on a single resource

Reviewer: Headers are using `Installing` and `Install`. I would recommend choosing one to make the page consistent (same thing with `Update` vs `Updating`).

```
.\install-latest-extension.ps1 -Username <USERNAME> -Password <PASSWORD> -SubscriptionId <SUBSCRIPTION_ID> -ResourceGroup <RESOURCE_GROUP_NAME> -SiteName <SITE_NAME> -DDApiKey <DATADOG_API_KEY> -ExtensionVersion <EXTENSION_VERSION>
```

Replace `<EXTENSION_VERSION>` with the version of the extension you wish to install. For instance, `1.4.0`.

#### Install specific version on an entire resource group

To install a specific version for a resource group, follow the [instructions for updating the extension for a resource group](#powershell-resource-group) and add the `-ExtensionVersion` parameter to the installation command.

```
.\update-all-site-extensions.ps1 -SubscriptionId <SUBSCRIPTION_ID> -ResourceGroup <RESOURCE_GROUP_NAME> -Username <USERNAME> -Password <PASSWORD> -ExtensionVersion <EXTENSION_VERSION>
```

Replace `<EXTENSION_VERSION>` with the version of the extension you wish to install. For instance, `1.4.0`.

## Troubleshooting

### If you are missing metrics and metadata in the APM trace panel and service page

> It is likely that you do not have the Azure integration configured to monitor your application. Proper configuration improves your ability to correlate metrics, traces, and logs in the Datadog platform, as well as to avoid broken or empty screens. To fix this:

Reviewer: ```suggestion
It is likely that you do not have the Azure integration configured to monitor your application. Proper configuration improves your ability to correlate metrics, traces, and logs on the Datadog site, as well as to avoid broken or empty screens. To fix this:
```

1. Ensure you have installed the [Azure integration][9] for the Azure subscription where your application is running.

> 2. Ensure that any App Service Plan filtering rules you may have applied include the App Service Plan where the app is running. Tags on the app itself are not used for filtering.

Reviewer: ```suggestion
2. Ensure that any App Service Plan filtering rules you applied include the App Service Plan where the app is running. Tags on the app itself are not used for filtering.
```

Learn more about [custom metrics][17].

## Extension management with Powershell

> Datadog provides automatic scripts with which you can update or install the Azure App Service Extension using Powershell. Scripted extension management enables you to [update extensions in bulk by resource group](#powershell-resource-group) and to [designate the installation of specific versions of the site extension](#powershell-specific-version). You can also use scripts to programatically add the extension in CI/CD pipelines, as well as discover and update extensions that are already installed.

Reviewer 2: I don't think the word "automatic" is needed here.

a. Verify you've set `DD_SITE` and `DD_API_KEY` correctly.
b. Do a full stop and start of your application.
c. If not resolved, try uninstalling the extension and re-installing (this also ensures you are running the latest version).

> d. Still need help? Contact [Datadog support][24].

Reviewer 3: The formatting of this section is buggy, but I think it was before you added a letter to the list. Lines 223-226 should be indented to line up with the "I" in "If" of line 213.
